File: tests/test-output-schema.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { describe, it } from "node:test";
import { fileURLToPath } from "node:url";
import {
	allow,
	deny,
	genericBlock,
	genericWarn,
	postWarn,
	stopBlock,
	stopWarn,
	warn,
} from "../hooks/scripts/_lib.mjs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const SCHEMA_PATH = resolve(
	__dirname,
	"..",
	"hooks",
	"schema",
	"hook-output.json",
);

let Ajv;
try {
	Ajv = (await import("ajv/dist/2020.js")).default;
} catch {
	try {
		Ajv = (await import("ajv")).default;
	} catch {
		// ajv not installed; schema validation tests will be skipped
	}
}

function captureExit(fn, ...args) {
	let captured = "";
	const origWrite = process.stdout.write.bind(process.stdout);
	const origExit = process.exit.bind(process);

	process.stdout.write = (chunk) => {
		captured += chunk;
		return true;
	};
	process.exit = () => {
		throw new Error("__exit__");
	};

	try {
		fn(...args);
	} catch (err) {
		if (!err.message.includes("__exit__")) throw err;
	} finally {
		process.stdout.write = origWrite;
		process.exit = origExit;
	}

	return JSON.parse(captured.trim());
}

function loadSchema() {
	return JSON.parse(readFileSync(SCHEMA_PATH, "utf8"));
}

describe("PostWarn", () => {
	it("should have correct output structure", () => {
		const output = captureExit(postWarn, "test message");
		assert.equal(output.hookSpecificOutput.hookEventName, "PostToolUse");
		assert.equal(output.hookSpecificOutput.additionalContext, "test message");
	});

	it("should have no extra fields", () => {
		const output = captureExit(postWarn, "msg");
		const keys = Object.keys(output.hookSpecificOutput);
		assert.deepEqual(
			new Set(keys),
			new Set(["hookEventName", "additionalContext"]),
		);
	});

	it("should validate against schema", { skip: !Ajv }, () => {
		const ajv = new Ajv();
		const validate = ajv.compile(loadSchema());
		const output = captureExit(postWarn, "test");
		assert.ok(validate(output), JSON.stringify(validate.errors));
	});
});

describe("GenericWarn", () => {
	it("should have correct output structure", () => {
		const output = captureExit(genericWarn, "warning text");
		assert.equal(output.hookSpecificOutput.hookEventName, "PostToolUse");
		assert.equal(output.hookSpecificOutput.additionalContext, "warning text");
	});

	it("should validate against schema", { skip: !Ajv }, () => {
		const ajv = new Ajv();
		const validate = ajv.compile(loadSchema());
		const output = captureExit(genericWarn, "test");
		assert.ok(validate(output), JSON.stringify(validate.errors));
	});
});

describe("GenericBlock", () => {
	it("should have correct output structure", () => {
		const output = captureExit(genericBlock, "blocked");
		assert.equal(output.hookSpecificOutput.hookEventName, "PreToolUse");
		assert.equal(output.hookSpecificOutput.permissionDecision, "deny");
		assert.equal(output.hookSpecificOutput.permissionDecisionReason, "blocked");
	});

	it("should validate against schema", { skip: !Ajv }, () => {
		const ajv = new Ajv();
		const validate = ajv.compile(loadSchema());
		const output = captureExit(genericBlock, "test");
		assert.ok(validate(output), JSON.stringify(validate.errors));
	});
});

describe("Deny", () => {
	it("should produce PreToolUse output", () => {
		const output = captureExit(deny, "reason");
		const hso = output.hookSpecificOutput;
		assert.equal(hso.hookEventName, "PreToolUse");
		assert.equal(hso.permissionDecision, "deny");
		assert.equal(hso.permissionDecisionReason, "reason");
	});

	it("should validate against schema", { skip: !Ajv }, () => {
		const ajv = new Ajv();
		const validate = ajv.compile(loadSchema());
		const output = captureExit(deny, "reason");
		assert.ok(validate(output), JSON.stringify(validate.errors));
	});
});

describe("Warn", () => {
	it("should default to PostToolUse event", () => {
		const output = captureExit(warn, "msg");
		assert.equal(output.hookSpecificOutput.hookEventName, "PostToolUse");
	});

	it("should validate against schema", { skip: !Ajv }, () => {
		const ajv = new Ajv();
		const validate = ajv.compile(loadSchema());
		const output = captureExit(warn, "msg");
		assert.ok(validate(output), JSON.stringify(validate.errors));
	});
});

describe("Allow", () => {
	it("should have correct output structure", () => {
		const output = captureExit(allow, "allowed");
		const hso = output.hookSpecificOutput;
		assert.equal(hso.hookEventName, "PreToolUse");
		assert.equal(hso.permissionDecision, "allow");
	});

	it("should validate against schema", { skip: !Ajv }, () => {
		const ajv = new Ajv();
		const validate = ajv.compile(loadSchema());
		const output = captureExit(allow, "allowed");
		assert.ok(validate(output), JSON.stringify(validate.errors));
	});
});

describe("StopWarn", () => {
	it("should write JSON to stdout with decision warn", () => {
		let captured = "";
		let exitCalled = false;
		const origWrite = process.stdout.write.bind(process.stdout);
		const origExit = process.exit;
		process.stdout.write = (chunk) => {
			captured += chunk;
			return true;
		};
		process.exit = () => {
			exitCalled = true;
			throw new Error("__exit__");
		};
		try {
			stopWarn("stale session");
		} catch (err) {
			if (!err.message.includes("__exit__")) throw err;
		} finally {
			process.stdout.write = origWrite;
			process.exit = origExit;
		}
		assert.ok(exitCalled, "stopWarn must call process.exit");
		const parsed = JSON.parse(captured.trim());
		assert.strictEqual(parsed.decision, "warn");
		assert.ok(parsed.reason.includes("stale session"));
	});
});

describe("StopBlock", () => {
	it("should exit with code 2 and write flat decision JSON to stdout", () => {
		let stdoutMsg = "";
		let exitCode = null;
		const origStderr = process.stderr.write.bind(process.stderr);
		const origStdout = process.stdout.write.bind(process.stdout);
		const origExit = process.exit.bind(process);
		process.stderr.write = (chunk) => {
			stderrMsg += chunk;
			return true;
		};
		process.stdout.write = (chunk) => {
			stdoutMsg += chunk;
			return true;
		};
		process.exit = (code) => {
			exitCode = code;
			throw new Error("__exit__");
		};
		try {
			stopBlock("placeholders found");
		} catch (err) {
			if (!err.message.includes("__exit__")) throw err;
		} finally {
			process.stderr.write = origStderr;
			process.stdout.write = origStdout;
			process.exit = origExit;
		}
		assert.equal(exitCode, 2);
		const parsed = JSON.parse(stdoutMsg.trim());
		assert.strictEqual(parsed.decision, "block");
		assert.ok(parsed.reason.includes("placeholders found"));
		assert.strictEqual(
			parsed.hookSpecificOutput,
			undefined,
			"stopBlock must not use hookSpecificOutput wrapper",
		);
	});
});

File: tests/test_output_schema.py
"""Tests that _lib output helpers produce valid CC hook JSON schema."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Callable

import pytest  # type: ignore[import-untyped]

sys.path.insert(0, str(Path(__file__).parent.parent / "hooks" / "scripts"))

SCHEMA_PATH = Path(__file__).parent.parent / "hooks" / "schema" / "hook-output.json"

HAS_JSONSCHEMA: bool = False
try:
    import jsonschema  # type: ignore[import-untyped]

    HAS_JSONSCHEMA = True
except ImportError:
    pass


def load_schema() -> dict[str, Any]:
    return json.loads(SCHEMA_PATH.read_text())  # type: ignore[no-any-return]


def run_and_capture(
    func: Callable[..., Any], *args: Any, capsys: pytest.CaptureFixture[str]
) -> dict[str, Any]:
    """Call a _lib output function, capture stdout, parse JSON."""
    with pytest.raises(SystemExit):
        func(*args)
    captured = capsys.readouterr()
    return json.loads(captured.out)  # type: ignore[no-any-return]


class TestPostWarn:
    def test_output_structure(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import post_warn  # type: ignore[import-not-found]

        output = run_and_capture(post_warn, "test message", capsys=capsys)
        assert output["hookSpecificOutput"]["hookEventName"] == "PostToolUse"
        assert output["hookSpecificOutput"]["additionalContext"] == "test message"

    def test_no_extra_fields(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import post_warn  # type: ignore[import-not-found]

        output = run_and_capture(post_warn, "msg", capsys=capsys)
        hso: dict[str, Any] = output["hookSpecificOutput"]
        assert set(hso.keys()) == {"hookEventName", "additionalContext"}

    @pytest.mark.skipif(not HAS_JSONSCHEMA, reason="jsonschema not installed")  # type: ignore[misc]
    def test_validates_against_schema(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import post_warn  # type: ignore[import-not-found]

        output = run_and_capture(post_warn, "test", capsys=capsys)
        jsonschema.validate(output, load_schema())


class TestGenericWarn:
    def test_output_structure(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import generic_warn  # type: ignore[import-not-found]

        output = run_and_capture(generic_warn, "warning text", capsys=capsys)
        assert output["reason"] == "warning text"
        assert "hookSpecificOutput" not in output

    @pytest.mark.skipif(not HAS_JSONSCHEMA, reason="jsonschema not installed")  # type: ignore[misc]
    def test_validates_against_schema(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import generic_warn  # type: ignore[import-not-found]

        output = run_and_capture(generic_warn, "test", capsys=capsys)
        jsonschema.validate(output, load_schema())


class TestGenericBlock:
    def test_output_structure(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import generic_block  # type: ignore[import-not-found]

        output = run_and_capture(generic_block, "blocked", capsys=capsys)
        assert output["decision"] == "block"
        assert output["reason"] == "blocked"
        assert "hookSpecificOutput" not in output

    @pytest.mark.skipif(not HAS_JSONSCHEMA, reason="jsonschema not installed")  # type: ignore[misc]
    def test_validates_against_schema(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import generic_block  # type: ignore[import-not-found]

        output = run_and_capture(generic_block, "test", capsys=capsys)
        jsonschema.validate(output, load_schema())


class TestDeny:
    """Verify deny() is only valid for PreToolUse context."""

    def test_pretooluse_output(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import deny  # type: ignore[import-not-found]

        output = run_and_capture(deny, "reason", capsys=capsys)
        hso: dict[str, Any] = output["hookSpecificOutput"]
        assert hso["hookEventName"] == "PreToolUse"
        assert hso["permissionDecision"] == "deny"
        assert hso["permissionDecisionReason"] == "reason"

    @pytest.mark.skipif(not HAS_JSONSCHEMA, reason="jsonschema not installed")  # type: ignore[misc]
    def test_validates_against_schema(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import deny  # type: ignore[import-not-found]

        output = run_and_capture(deny, "reason", capsys=capsys)
        jsonschema.validate(output, load_schema())


class TestWarn:
    """Verify warn() default event is PostToolUse."""

    def test_default_event(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import warn  # type: ignore[import-not-found]

        output = run_and_capture(warn, "msg", capsys=capsys)
        assert output["hookSpecificOutput"]["hookEventName"] == "PostToolUse"

    @pytest.mark.skipif(not HAS_JSONSCHEMA, reason="jsonschema not installed")  # type: ignore[misc]
    def test_validates_against_schema(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import warn  # type: ignore[import-not-found]

        output = run_and_capture(warn, "msg", capsys=capsys)
        jsonschema.validate(output, load_schema())


class TestAllow:
    def test_output_structure(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import allow  # type: ignore[import-not-found]

        output = run_and_capture(allow, "allowed", capsys=capsys)
        hso: dict[str, Any] = output["hookSpecificOutput"]
        assert hso["hookEventName"] == "PreToolUse"
        assert hso["permissionDecision"] == "allow"

    @pytest.mark.skipif(not HAS_JSONSCHEMA, reason="jsonschema not installed")  # type: ignore[misc]
    def test_validates_against_schema(self, capsys: pytest.CaptureFixture[str]) -> None:
        from _lib import allow  # type: ignore[import-not-found]

        output = run_and_capture(allow, "allowed", capsys=capsys)
        jsonschema.validate(output, load_schema())
